### ISSUE TITLE
Implement Auto model Selection and Best Fit.

### ISSUE THREAD

**Issue by sevin7352** [sevin7352/BlackJackTrainner#6]
When Retraining use same data model.
When Using a fresh Model, Test Multiple models and settings for those models.
only Save Model if it performs better than pervious models

From ChatGpt for advice on how to structure it.
To test different models and settings at once and find the best fit for your problem, you can employ model selection and hyperparameter tuning techniques. Here's how you can approach this:

1. Manual Testing with Different Models
You can create different pipelines with varying models and hyperparameters, then evaluate them based on performance metrics such as accuracy, precision, recall, etc.

Step-by-Step Approach:
Define multiple pipelines: Use different algorithms, such as SDCA, FastTree, or LightGBM.
Evaluate each model: Use a validation dataset to check performance.
Compare metrics: Track metrics like accuracy or AUC and choose the best performing model.
Example: Comparing Multiple Models
csharp
Copy
Edit
public void CompareModels(List<NueralNetTrainingDataEntry> data)
{
    // Load data into IDataView format
    var trainData = mlContext.Data.LoadFromEnumerable(data);
    
    // Split the data into training and validation sets
    var trainTestData = mlContext.Data.TrainTestSplit(trainData, testFraction: 0.2);

    // Define multiple models to compare
    var models = new List<ITransformer>
    {
        // SDCA Model
        TrainSdcaModel(trainTestData.TrainSet),
        
        // FastTree Model
        TrainFastTreeModel(trainTestData.TrainSet),
        
        // LightGBM Model
        TrainLightGbmModel(trainTestData.TrainSet)
    };

    // Evaluate each model
    foreach (var model in models)
    {
        var predictions = model.Transform(trainTestData.TestSet);
        var metrics = mlContext.MulticlassClassification.Evaluate(predictions, labelColumnName: "Action");

        Console.WriteLine($"Model: {model.GetType().Name}");
        Console.WriteLine($"Micro Accuracy: {metrics.MicroAccuracy:P2}");
        Console.WriteLine($"Macro Accuracy: {metrics.MacroAccuracy:P2}");
        Console.WriteLine($"Log-loss: {metrics.LogLoss:P2}");
    }
}

private ITransformer TrainSdcaModel(IDataView trainData)
{
    var pipeline = mlContext.Transforms.Conversion.MapValueToKey("Action")
        .Append(mlContext.Transforms.Concatenate("Features", "PlayerHandSum", "DealerUpCard", "NumberOfAces", "CanSplit"))
        .Append(mlContext.MulticlassClassification.Trainers.SdcaMaximumEntropy("Action", "Features"))
        .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedAction", "PredictedLabel"));
    return pipeline.Fit(trainData);
}

private ITransformer TrainFastTreeModel(IDataView trainData)
{
    var pipeline = mlContext.Transforms.Conversion.MapValueToKey("Action")
        .Append(mlContext.Transforms.Concatenate("Features", "PlayerHandSum", "DealerUpCard", "NumberOfAces", "CanSplit"))
        .Append(mlContext.MulticlassClassification.Trainers.FastTree("Action", "Features"))
        .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedAction", "PredictedLabel"));
    return pipeline.Fit(trainData);
}

private ITransformer TrainLightGbmModel(IDataView trainData)
{
    var pipeline = mlContext.Transforms.Conversion.MapValueToKey("Action")
        .Append(mlContext.Transforms.Concatenate("Features", "PlayerHandSum", "DealerUpCard", "NumberOfAces", "CanSplit"))
        .Append(mlContext.MulticlassClassification.Trainers.LightGbm("Action", "Features"))
        .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedAction", "PredictedLabel"));
    return pipeline.Fit(trainData);
}
2. Hyperparameter Tuning with Cross-Validation
To optimize the settings for each model (like the number of iterations, regularization strength, etc.), you can use hyperparameter tuning.

Using Hyperparameter Tuning with Cross-Validation:
Define a range of values for each hyperparameter (e.g., maximum iterations, learning rate).
Run cross-validation to evaluate different combinations of hyperparameters.
Select the best combination based on performance metrics.
ML.NET doesn't have built-in automated hyperparameter tuning like some other frameworks (e.g., GridSearchCV in Python), but you can manually implement it by looping through different settings.

Example: Manually Tuning Hyperparameters
csharp
Copy
Edit
public void TuneHyperparameters(List<NueralNetTrainingDataEntry> data)
{
    var trainData = mlContext.Data.LoadFromEnumerable(data);
    var trainTestData = mlContext.Data.TrainTestSplit(trainData, testFraction: 0.2);
    
    // Define hyperparameter grids
    var iterationsOptions = new[] { 10, 20, 50 };
    var lambdaOptions = new[] { 0.1, 0.5, 1.0 };

    double bestAccuracy = 0.0;
    ITransformer bestModel = null;

    foreach (var iterations in iterationsOptions)
    {
        foreach (var lambda in lambdaOptions)
        {
            var model = TrainSdcaModelWithHyperparameters(trainTestData.TrainSet, iterations, lambda);
            var predictions = model.Transform(trainTestData.TestSet);
            var metrics = mlContext.MulticlassClassification.Evaluate(predictions, labelColumnName: "Action");

            Console.WriteLine($"Iterations: {iterations}, Lambda: {lambda}, Micro Accuracy: {metrics.MicroAccuracy:P2}");

            if (metrics.MicroAccuracy > bestAccuracy)
            {
                bestAccuracy = metrics.MicroAccuracy;
                bestModel = model;
            }
        }
    }

    Console.WriteLine($"Best Accuracy: {bestAccuracy:P2}");
    SaveModel(bestModel, "best_model.zip");
}

private ITransformer TrainSdcaModelWithHyperparameters(IDataView trainData, int maxIterations, double lambda)
{
    var pipeline = mlContext.Transforms.Conversion.MapValueToKey("Action")
        .Append(mlContext.Transforms.Concatenate("Features", "PlayerHandSum", "DealerUpCard", "NumberOfAces", "CanSplit"))
        .Append(mlContext.MulticlassClassification.Trainers.SdcaMaximumEntropy("Action", "Features", maximumNumberOfIterations: maxIterations, l2Regularization: lambda))
        .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedAction", "PredictedLabel"));
    
    return pipeline.Fit(trainData);
}
3. Grid Search for Multiple Hyperparameters
If you want to automate the search, you could implement a grid search approach, trying all combinations of hyperparameters. While ML.NET doesn't have a built-in grid search, you can build it manually by looping through possible values.

csharp
Copy
Edit
var maxIterationsValues = new int[] { 10, 20, 50 };
var lambdaValues = new double[] { 0.1, 0.5, 1.0 };
4. Use Cross-Validation for More Reliable Results
To prevent overfitting to your training data, you should consider cross-validation. ML.NET supports K-fold cross-validation to better assess model performance.

csharp
Copy
Edit
var crossValidationResults = mlContext.MulticlassClassification.CrossValidate(data: trainData, 
    estimator: pipeline, numFolds: 5, labelColumnName: "Action");

foreach (var result in crossValidationResults)
{
    Console.WriteLine($"Fold: {result.Fold}, Accuracy: {resultMetrics.MicroAccuracy}");
}
5. Automated Model Selection (ML.NET doesn't have this out-of-the-box)
ML.NET doesn't directly support automated hyperparameter tuning and model selection like libraries in Python (e.g., Scikit-learn).
However, by using cross-validation and manual hyperparameter tuning, you can still achieve good results.
🎯 Key Takeaways:
Test multiple models and hyperparameters manually by comparing metrics like accuracy, precision, recall, or AUC.
Use cross-validation for more robust model selection.
Optimize hyperparameters through a grid search or manual tuning to find the best combination.
This method will allow you to determine which model and hyperparameters work best for your data! Let me know if you need further clarification on any part of this. 😊
